### PR TITLE
[SYCLomatic] Fix the __constant__ variable migration

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 1]
```diff
@@ -9480,6 +9480,33 @@ bool MemVarRule::currentIsHost(const VarDecl *VD, std::string VarName) {
...
+      if (BO) {
+        if (BO->isAssignmentOp() && (BO->getRHS() == Current.get<Expr>()) &&
+            BO->getLHS()->getType()->isPointerType() &&
+            !BO->getLHS()->getType()->getPointeeType().isConstQualified())
+          return BO->getRHS();
+      } else if (VD) {
+        if (VD->hasInit() && (VD->getInit() == Current.get<Expr>()) &&
+            VD->getType()->isPointerType() &&
+            !VD->getType()->getPointeeType().isConstQualified())
+          return VD->getInit();
```
Comment (zhiweij1) [inline clang/lib/DPCT/ASTTraversal.cpp:9504]:
```suggestion
          return VD->getInit();
        return nullptr;
```

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 2]
```diff
@@ -9480,6 +9480,33 @@ bool MemVarRule::currentIsHost(const VarDecl *VD, std::string VarName) {
...
+      }
```
Comment (zhiweij1) [inline clang/lib/DPCT/ASTTraversal.cpp:9503]:
```suggestion
      } else {
        return nullptr;
      }
```

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 3]
```diff
@@ -9480,6 +9480,33 @@ bool MemVarRule::currentIsHost(const VarDecl *VD, std::string VarName) {
 }
 
 void MemVarRule::runRule(const MatchFinder::MatchResult &Result) {
+  auto getRHSOfTheNonConstAssignedVar = [](const DeclRefExpr *DRE,
+                                           const Decl *Range) -> const Expr * {
+    auto &Context = DpctGlobalInfo::getContext();
+    DynTypedNode Current = DynTypedNode::create(*DRE);
+    DynTypedNodeList Parents = Context.getParents(Current);
+    while (!Parents.empty()) {
+      if (Parents[0].get<Decl>() && Parents[0].get<Decl>() == Range)
```
Comment (zhiweij1) [inline clang/lib/DPCT/ASTTraversal.cpp]:
```suggestion

```

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 4]
```diff
@@ -9480,6 +9480,33 @@ bool MemVarRule::currentIsHost(const VarDecl *VD, std::string VarName) {
...
+        break;
+      const BinaryOperator *BO = Parents[0].get<BinaryOperator>();
+      const VarDecl *VD = Parents[0].get<VarDecl>();
+      if (BO) {
+        if (BO->isAssignmentOp() && (BO->getRHS() == Current.get<Expr>()) &&
```
Comment (zhiweij1) [inline clang/lib/DPCT/ASTTraversal.cpp]: 9494 and 9499 can use a function.

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 5]
```diff
@@ -9543,6 +9570,7 @@ void MemVarRule::runRule(const MatchFinder::MatchResult &Result) {
       return std::make_pair(Begin, Length);
     };
     const auto *Parent = getParentStmt(MemVarRef);
+    bool ReplacementEmplaced = false;
```
Comment (zhiweij1) [inline clang/lib/DPCT/ASTTraversal.cpp]:
```suggestion
    bool HasTypeCasted = false;
```

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 6]
```diff
@@ -9577,9 +9606,23 @@ void MemVarRule::runRule(const MatchFinder::MatchResult &Result) {
...
           }
+          ReplacementEmplaced = true;
         }
       }
     }
+    if (!ReplacementEmplaced && Decl->hasAttr<CUDAConstantAttr>() &&
+        (MemVarRef->getType()->getTypeClass() ==
+             Type::TypeClass::ConstantArray ||
+         MemVarRef->getType()->getTypeClass() == Type::TypeClass::Pointer)) {
+      const Expr *RHS = getRHSOfTheNonConstAssignedVar(MemVarRef, Func);
```
Comment (zhiweij1) [inline clang/lib/DPCT/ASTTraversal.cpp]:
```suggestion
      const Expr *RHS = getRHSOfTheNonConstAssignedVar(MemVarRef);
```

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 7]
```diff
@@ -9577,9 +9606,23 @@ void MemVarRule::runRule(const MatchFinder::MatchResult &Result) {
...
                                             ">(&", Decl->getName(), ")")));
           }
+          ReplacementEmplaced = true;
         }
       }
     }
+    if (!ReplacementEmplaced && Decl->hasAttr<CUDAConstantAttr>() &&
+        (MemVarRef->getType()->getTypeClass() ==
+             Type::TypeClass::ConstantArray ||
+         MemVarRef->getType()->getTypeClass() == Type::TypeClass::Pointer)) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/ASTTraversal.cpp]: This condition can be removed. Test case kernel6 is caused by `&`, it should not be processed in this `if` block

File: clang/lib/DPCT/ASTTraversal.cpp (hunk 3 above)
Comment (zhiweij1) [reply clang/lib/DPCT/ASTTraversal.cpp]: done

File: clang/lib/DPCT/ASTTraversal.cpp (hunk 1 above)
Comment (zhiweij1) [reply clang/lib/DPCT/ASTTraversal.cpp:9504]: done

File: clang/lib/DPCT/ASTTraversal.cpp (hunk 2 above)
Comment (zhiweij1) [reply clang/lib/DPCT/ASTTraversal.cpp:9503]: done

File: clang/lib/DPCT/ASTTraversal.cpp (hunk 5 above)
Comment (zhiweij1) [reply clang/lib/DPCT/ASTTraversal.cpp]: done

File: clang/lib/DPCT/ASTTraversal.cpp (hunk 6 above)
Comment (zhiweij1) [reply clang/lib/DPCT/ASTTraversal.cpp]: done

File: clang/lib/DPCT/ASTTraversal.cpp (hunk 4 above)
Comment (zhiweij1) [reply clang/lib/DPCT/ASTTraversal.cpp]: done

File: clang/lib/DPCT/ASTTraversal.cpp (hunk 7 above)
Comment (zhiweij1) [reply clang/lib/DPCT/ASTTraversal.cpp]: done

**Review by ziranzha — APPROVED** [review]

**Review by tomflinda — APPROVED** [review]
LGTM

**Pull request merged**